src/node_large_page.cc

```diff
@@ -0,0 +1,253 @@
+// Copyright (C) 2018 Intel Corporation
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom
+// the Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+// OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
+// OR OTHER DEALINGS IN THE SOFTWARE.
+//
+// SPDX-License-Identifier: MIT
+
+#include <errno.h>
+#include <fcntl.h>  // _O_RDWR
+#include <limits.h>  // PATH_MAX
+#include <locale.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/mman.h>
+
+#include <string>
+#include <fstream>
+#include <iostream>
+#include <sstream>
+
+// The functions in this file map the text segment of node into 2M pages.
+// The algorithm is simple
+// 1. Find the text region of node binary in memory
+//    Examine the /proc/self/maps to determine the currently mapped text
+//    region and obtain the start and end
+//    Modify the start to point to the very beginning of node text segment
+//    (from variable nodetext setup in ld.script)
+//    Align the address of start and end to Large Page Boundaries
+//
+// 2. Move the text region to large pages
+//    Map a new area and copy the original code there
+//    Use mmap using the start address with MAP_FIXED so we get exactly the
+//      same virtual address
+//    Use madvise with MADV_HUGE_PAGE to use Anonymous 2M Pages
+//    If successful copy the code there and unmap the original region.
+
+extern char __executable_start;
+extern char __etext;
+extern char __nodetext;
+
+namespace node {
+#define ALIGN(x, a)           (((x) + (a) - 1) & ~((a) - 1))
+#define PAGE_ALIGN_UP(x, a)   ALIGN(x, a)
+#define PAGE_ALIGN_DOWN(x, a) ((x) & ~((a) - 1))
```

Comment:
style nit: these could all be inline functions, right?

Reply:
Thanks. Yes fixed them to be inline functions.